src/psyclone/tests/psyad/adjoint_visitor_test.py

```diff
@@ -413,6 +443,151 @@ def test_assignment_node(fortran_reader, fortran_writer):
+    if no active variables are found.
+
+    '''
+    tl_psyir = fortran_reader.psyir_from_source(TL_LOOP_CODE)
+    loop = tl_psyir.walk(Loop)[0]
+    assert isinstance(loop, Loop)
+    adj_visitor = AdjointVisitor(["a", "b", "c"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor.loop_node(loop)
+    assert ("A loop node should not be visited before a schedule, as "
+            "the latter sets up the active variables." in str(info.value))
+
+
+def test_loop_node_bounds_error(fortran_reader):
+    '''Test that the loop_node method raises the expected exception if an
+    active variable is found in a loop bound (lower, upper or step) or
+    the iterator is an active variable.
+
+    '''
+    tl_psyir = fortran_reader.psyir_from_source(TL_LOOP_CODE)
+    # lower bound
+    adj_visitor = AdjointVisitor(["a", "b", "c", "lo"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor(tl_psyir)
+    assert ("The lower bound of a loop should not contain active variables, "
+            "but found 'lo'" in str(info.value))
+    # upper bound
+    adj_visitor = AdjointVisitor(["a", "b", "c", "hi"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor(tl_psyir)
+    assert ("The upper bound of a loop should not contain active variables, "
+            "but found 'hi'" in str(info.value))
+    # step
+    adj_visitor = AdjointVisitor(["a", "b", "c", "step"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor(tl_psyir)
+    assert ("The step of a loop should not contain active variables, "
+            "but found 'step'" in str(info.value))
+    # loop variable
+    adj_visitor = AdjointVisitor(["a", "b", "c", "i"])
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor(tl_psyir)
+    assert ("The loop iterator 'i' should not be an active variable."
+            in str(info.value))
+
+
+def test_loop_node_passive(fortran_reader):
+    '''Test that the loop_node method raises an exception if there are no
+    active variables within the supplied loop node. This is because
+    the schedule node should have already dealt with this case.
+
+    '''
+    tl_psyir = fortran_reader.psyir_from_source(TL_LOOP_CODE)
+    tl_loop = tl_psyir.walk(Loop)[0]
+    assert isinstance(tl_loop, Loop)
+    adj_visitor = AdjointVisitor(["d", "e"])
+    ad_psyir = adj_visitor._visit(tl_psyir)
+    ad_loop = ad_psyir.walk(Loop)[0]
+    assert isinstance(ad_loop, Loop)
+
+    with pytest.raises(VisitorError) as info:
+        _ = adj_visitor.loop_node(tl_loop)
+    assert ("A passive loop node should not be processed by the loop_node() "
+            "method within the AdjointVisitor() class, as it should have been "
+            "dealt with by the schedule_node() method." in str(info.value))
+
+
+@pytest.mark.parametrize("in_bounds,out_bounds", [
+    ("lo,hi", "hi, lo, -1"),
+    ("lo,hi,1", "hi, lo, -1"),
+    ("lo,hi,-1", "hi - MOD(hi - lo, - 1), lo, 1"),
```

Comment:
I thought the code checked and if the increment was +/-1 then no offset was required?

Reply:
Yes, me too, and it does if the literal value is -1. However, in this case -1 is represented in the PSyIR as a unary minus containing a literal with value 1, due to the way it is originally stored in Fortran. I could check for this as well but decided not to as I think that having -1 as a step is pretty rare.